EIPS/eip-758.md

```diff
@@ -0,0 +1,99 @@
+
+## Motivation
+External callers presently have no way of accessing return data from Ethereum, if the function was executed via `eth_sendTransaction` or `eth_sendRawTransaction` RPC request.  Access to function return data is in many cases a desirable feature.  Making return data available to external callers also addresses the inconsistency between internal callers, which have access to return data within the context of the transaction, and external callers, which do not.  Presently, a common workaround is to log the return data, which is bad for several reasons: it contributes to chain bloat, imposes additional gas costs on the caller, and can result in unused logs being written if the externally called function involves other (internal) function calls that log their return data.
+
+## Specification
+
+### Subscription
+A caller who wants to be notified of return data for their transactions sends an `eth_subscribe` RPC request with the parameter `"returnData"`:
+
+```json
+{"jsonrpc": "2.0", "id": 1, "method": "eth_subscribe", "params": ["returnData"]}
+```
+
+The Ethereum node responds with a subscription ID:
+
+```json
+{"jsonrpc": "2.0", "id": 1, "result": "0x00000000000000000000000000000b0b"}
+```
+
+The caller submits a transaction via `eth_sendTransaction` or `eth_sendRawTransaction` RPC request which has the transaction hash `"0x00000000000000000000000000000000000000000000000000000000deadbeef"`.  When the transaction is sealed (mined), the Ethereum node computes the return value (`"0x000000000000000000000000000000000000000000000000000000000000002a"`) and pushes a notification to the caller:
+
+```json
+{
+  "jsonrpc": "2.0",
+  "method": "eth_subscription",
+  "params": {
+    "result": {
+      "transactionHash": "0x00000000000000000000000000000000000000000000000000000000deadbeef",
+      "returnData": "0x000000000000000000000000000000000000000000000000000000000000002a"
+    },
+    "subscription": "0x00000000000000000000000000000b0b"
+  }
+}
+```
+
+The caller receives notifications about their transactions' return data in two cases: first when a transaction is sealed, and again (with an extra `"removed": true` field) if a transaction is affected by a chain reorganization.  Notifications are sent to the client for all transactions submitted from the client that are sealed _after_ subscribing.  As with other subscriptions, the caller can send an `eth_unsubscribe` RPC request to stop receiving push notifications:
+
+```json
+{"jsonrpc": "2.0", "id": 2, "method": "eth_unsubscribe", "params": ["0x00000000000000000000000000000b0b"]}
+```
```

Comment:
I'm behaving a bit out of EIP editorship, but I think usually asynchronous methods take callback functions as arguments: for example
https://github.com/ethereum/wiki/wiki/JavaScript-API#web3ethfilter

Maybe a similar interface is easier to implement.

Reply:
I may be misunderstanding your comment, but, the asynchronous callback is defined in the RPC library (web3.js, in your link).  The code snippets I included are intended to be the actual JSONs that go out over the wire to the Ethereum node.  For example, using curl to send an `eth_newFilter` request (to a Rinkeby node, in this case):

```sh
curl -H "Content-Type: application/json" -X POST --data '{"jsonrpc": "2.0", "id": 1, "method": "eth_newFilter", "params": ["returnData"]}' http://rinkeby.ethereum.nodes.augur.net
```

web3.js's `eth.filter` function is a JavaScript wrapper around `eth_newFilter`.  This wrapper function accepts a callback (which fires any time it sees events matching the filter), but the callback is just a JavaScript function that is held in memory by web3.js.  The Ethereum node never knows anything about the callback function!

(See the wiki pages for [subscriptions](https://github.com/ethereum/go-ethereum/wiki/RPC-PUB-SUB) and [filters](https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_newfilter) for more examples and better explanations.)

Reply:
Thank you for the reference.  Right, callback functions shouldn't be sent around on wire.